sway/tree/workspace.c

```diff
@@ -513,3 +514,113 @@ struct sway_container *workspace_output_get_highest_available(
 
 	return NULL;
 }
+
+struct pid_workspace {
+	pid_t pid;
+	char *workspace;
+	struct timespec time_added;
+
+	struct sway_container *output;
+	struct wl_listener output_destroy;
+
+	struct wl_list link;
+};
+
+static struct wl_list pid_workspaces;
+
+struct sway_container *workspace_for_pid(pid_t pid) {
+	if (!pid_workspaces.prev && !pid_workspaces.next) {
+		wl_list_init(&pid_workspaces);
+		return NULL;
+	}
+
+	struct sway_container *ws = NULL;
+	struct pid_workspace *pw = NULL;
+
+	wlr_log(L_DEBUG, "Looking up workspace for pid %d", pid);
+
+	do {
+		struct pid_workspace *_pw = NULL;
+		wl_list_for_each(pw, &pid_workspaces, link) {
+			if (pid == pw->pid) {
+				wlr_log(L_DEBUG,
+						"found pid_workspace for pid %d, workspace %s",
+						pid, pw->workspace);
+				_pw = pw;
+				break;
+			}
+		}
+
+		if (_pw) {
```

Comment:
I think using a goto to break out of a nested loop is cleaner than this.

Reply:
It's not just to break out of the loop. If we don't find a pid_workspace at the end of wl_list_for_each, pw will still be set to the last one in the list.

Reply:
I can still simplify this nonetheless. New patch coming up momentarily.